src/plugins/discover/public/application/main/hooks/use_discover_state.ts

```diff
@@ -246,26 +246,19 @@ export function useDiscoverState({
   const fetchResults = useCallback(() => {
     if (documentState.result?.length) {
       const firstRow = documentState.result[0];
-      const columns = Object.keys(firstRow.raw).slice(0, MAX_NUM_OF_COLUMNS);
-      if (!isEqual(columns, documentStateCols)) {
-        return Object.keys(firstRow.raw).slice(0, MAX_NUM_OF_COLUMNS);
-      }
-      return [];
+      return Object.keys(firstRow.raw).slice(0, MAX_NUM_OF_COLUMNS);
     }
     return [];
-  }, [documentState.result, documentStateCols]);
+  }, [documentState.result]);
 
   useEffect(() => {
     async function fetchDataview() {
-      if (state.query && isOfAggregateQueryType(state.query)) {
+      if (isPlainRecord(state.query) && documentState.fetchStatus === FetchStatus.COMPLETE) {
         const indexPatternFROMQuery = getIndexPatternFromSQLQuery(state.query.sql);
         const idsTitles = await indexPatterns.getIdsWithTitle();
         const dataViewObj = idsTitles.find(({ title }) => title === indexPatternFROMQuery);
         if (dataViewObj) {
           const columns = fetchResults();
-          if (columns.length) {
```

Comment:
This is super important. If you remove this then we have the following bug.
Change the columns order on the sql mode, hit refresh (or change the timepicker), the columns order changes again

Reply:
I've returned it.